### ISSUE TITLE
X-Registry-Auth with very long passwords produces malformed base64

### ISSUE THREAD

**Issue by wrouesnel** [docker/docker#16840]
I was using the following password for a private registry: (user = `root`)
```
iL\T2ORDom=21;5IGR?X*;.LS8cmzUaFzNzwK&l*s#z7E$|RiI|\i;'h0X%,;o=I
```
(and am not using it now) - but I found the between docker and docker-py, I couldn't login to the private registry using the latter.

From some investigation it looks like Docker's encoding of the password is wrong, but since it's wrong in both directions the auth still works - whereas because docker-py is in Python, the correct encoding doesn't work.

In tests with the docker socket, the following was the base64 auth string passed with `X-Registry-Auth` by the Go-docker client:
```
eyJ1c2VybmFtZSI6InJvb3QiLCJwYXNzd29yZCI6ImlMXFxUMk9SRG9tPTIxOzVJR1I_WCo7LkxTOGNtelVhRnpOendLXHUwMDI2bCpzI3o3RSR8UmlJfFxcaTsnaDBYJSw7bz1JIiwiYXV0aCI6IiIsImVtYWlsIjoidy5yb3Vlc25lbEBnbWFpbC5jb20iLCJzZXJ2ZXJhZGRyZXNzIjoiZG9ja2VyLmludHJhbmF2LmlvIn0=
```

Running that through this test decode in Go:
```go
package main

import (
	"fmt"
	"encoding/base64"
)

func main() {
	str := "eyJ1c2VybmFtZSI6InJvb3QiLCJwYXNzd29yZCI6ImlMXFxUMk9SRG9tPTIxOzVJR1I_WCo7LkxTOGNtelVhRnpOendLXHUwMDI2bCpzI3o3RSR8UmlJfFxcaTsnaDBYJSw7bz1JIiwiYXV0aCI6IiIsImVtYWlsIjoidy5yb3Vlc25lbEBnbWFpbC5jb20iLCJzZXJ2ZXJhZGRyZXNzIjoiZG9ja2VyLmludHJhbmF2LmlvIn0="
	
	d, _ := base64.StdEncoding.DecodeString(str)
	fmt.Printf(string(d))
}
```
produces
```
{"username":"root","password":"iL\\T2ORDom=21;5I
```
as output.

Running the same string through pythons base64 decode produces an "incorrect padding" error. If I tweak the padding a little I get the following:
```python
base64.b64decode("""eyJ1c2VybmFtZSI6InJvb3QiLCJwYXNzd29yZCI6ImlMXFxUMk9SRG9tPTIxOzVJR1I_WCo7LkxTOGNtelVhRnpOendLXHUwMDI2bCpzI3o3RSR8UmlJfFxcaTsnaDBYJSw7bz1JIiwiYXV0aCI6IiIsImVtYWlsIjoidy5yb3Vlc25lbEBnbWFpbC5jb20iLCJzZXJ2ZXJhZGRyZXNzIjoiZG9ja2VyLmludHJhbmF2LmlvIn0==""")
Out[16]: '{"username":"root","password":"iL\\\\T2ORDom=21;5IGR\x16\n\x8e\xcb\x93\x14\xce\x18\xdb^\x95XQ\x9e\x93\x9e\x9d\xd2\xd7\x1dL\x0c\x0c\x8d\x9b\n\x9c\xc8\xde\x8d\xd1I\x1f\x14\x9aR_\x17\x17\x1aN\xc9\xda\x0c\x16\tK\x0e\xdb\xcfRH\x8b\x08\x98]]\x1a\x08\x8e\x88\x88\x8b\x08\x99[XZ[\x08\x8e\x88\x9d\xcb\x9c\x9b\xddY\\\xdb\x99[\x10\x19\xdbXZ[\x0b\x98\xdb\xdbH\x8b\x08\x9c\xd9\\\x9d\x99\\\x98Y\x19\x1c\x99\\\xdc\xc8\x8e\x88\x99\x1b\xd8\xda\xd9\\\x8b\x9a[\x9d\x1c\x98[\x98]\x8b\x9a[\xc8\x9f'

```

From what I can see, this looks like Docker is either chopping a padding byte off (but I couldn't see any obvious locations) or there's a mix up of unicode processing which means long and / or complicated passwords produce bad json / base64 encoding for the header - but since it's all on the Go-side, superficially it continues to work - but it breaks inter-op with external tools.

I was able to fix the problem by switching to a shorter, alphanumeric only password in the mean time.

Docker version:
```
Docker version 1.8.2, build 0a8c2e3
```
Docker Info
```
Containers: 36
Images: 177
Storage Driver: zfs
 Zpool: hddpool
 Zpool Health: ONLINE
 Parent Dataset: hddpool/var/lib/docker
 Space Used By Parent: 13303734272
 Space Available: 253963046912
 Parent Quota: no
 Compression: zle
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.2.3-040203-generic
Operating System: Ubuntu 15.04
CPUs: 4
Total Memory: 7.413 GiB
Name: will-laptop
ID: LY6E:BHJN:H3LU:NMTP:PBEZ:EIHR:RJVQ:XQY6:3WLG:WJEB:ZH3O:4NUW
WARNING: No swap limit support
```
```
Linux will-laptop 4.2.3-040203-generic #201510030832 SMP Sat Oct 3 12:34:31 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by MHBauer** [comment]
Looks like something before the base64 encoding tries to escape the backslash.
```
      original: iL\T2ORDom=21;5I
after encoding: iL\\T2ORDom=21;5I
```
I don't know where the code that does this is.

**Comment by vikstrous** [comment]
In JWT tokens the `/` is replaced with `_` and the padding is removed. Docker hub uses that scheme and docker-py is using regular base64 encoding. This is a bug in docker-py. I'll open an issue there for you. This issue can be closed.

**Comment by jlhawn** [comment]
The CLI is using the URL encoding rather than the standard base64 encoding [1] and so is the server [2].

[1] https://github.com/docker/docker/blob/634a848b8e3bdd8aed834559f3b2e0dfc7f5ae3a/api/client/utils.go#L138
[2] https://github.com/docker/docker/blob/2d1414e8b0110b2f1857adcafbb65484d5c003fd/api/server/router/local/image.go#L96

docker-py should also use the URL encoding if it is not already. The docs are not specific about using either the URL-safe or standard base64 encoding (http://docs.docker.com/reference/api/docker_remote_api_v1.21/#create-an-image) and this should be clarified.

**Comment by jlhawn** [comment]
ping @shin-

**Comment by shin-** [comment]
Fixed in docker-py 1.5.0, which we just released today.

**Comment by thaJeztah** [comment]
Thanks @shin- !

And thanks for reporting @wrouesnel, nice find!
